// enconvo/llm/llm.js
import {ChatOpenAI} from "./openai";

module.exports = function (options) {

    let chat = undefined
    if (options.llm === "enconvo.azure-openai") {
        chat = new ChatOpenAI({
            streaming: options.stream ? options.stream : false,
            temperature: +options.temperature,
            azureOpenAIApiKey: options.apiKey,
            azureOpenAIApiVersion: options.apiVersion,
            azureOpenAIApiDeploymentName: options.model,
            azureOpenAIApiInstanceName: options.instanceName,
            verbose: options.verbose ? options.verbose : false
        });
    } else if (options.llm === "enconvo.chatgpt") {
        chat = new ChatOpenAI({
            modelName: options.model,
            streaming: options.stream ? options.stream : false,
            temperature: +options.temperature,
            openAIApiKey: options.apiKey,
            verbose: options.verbose ? options.verbose : false
        }, {
            basePath: `${options.basePath}/v1/`
        });
    } else if (options.llm === "ysnows.openrouter") {
        chat = new ChatOpenAI({
            modelName: options.model,
            streaming: options.stream ? options.stream : false,
            temperature: +options.temperature,
            openAIApiKey: options.apiKey,
            verbose: options.verbose ? options.verbose : false
        }, {
            basePath: `${options.basePath}/v1`,
            baseOptions: {
                headers: {
                    "Http-Referer": "https://enconvo.com",
                    "X-Title": "enconvo"
                }
            }
        });
    }
    return chat
}
